mgmt: hawkbit: change hawkbit interface to support directly supplying server ip address

nealjack opened the pull request:
The hawkbit subsystem assumes that `getaddrinfo` will supply a valid, usable IP address. This isn't the case when using a network layer that depends on NAT64 for IPv4 traffic, like OpenThread. A DNS address obtained while behind a NAT64 requires conversion to IPv6 and adding the NAT64 prefix. As of now, `getaddrinfo` is not aware of NAT64 and does not handle this. In this case it can be better for the application itself to handle DNS resolution and provide properly prefixed IP addresses to the hawkbit subsystem.

This pull request adds an additional interface field (`server_hostname`) for the server hostname and changes `server_addr` to optionally point to an IP address. This is a potentially less confusing naming scheme, as hawkbit _does not_ currently work if an address is provided to `server_addr` instead of a hostname.

This commit also increases the default DNS name length slightly to accommodate larger hostnames longer than 20 bytes, i.e. (hawkbit.yourcompany.com). I also added some error handling in case a hostname is too long, so the hawkbit interface returns an error instead of quietly using a truncated and incorrect hostname.

This commit also changes how the `controllerId` is generated based on device id, and disentangles the two. The `controllerId` is what hawkbit uses to uniquely identify a device, and is not necessarily the same as the device id, and should be fully customizable by the user if needed. Previously, all custom device ids were being prepended with `CONFIG_BOARD`. When a user selects `CONFIG_HAWKBIT_CUSTOM_DEVICE_ID`,  they should be able to specify the full `controllerId` used with hawkbit.

nealjack committed:
mgmt: hawkbit: interface support for ip addresses and domain name

Previously, hawkbit interface only supported a url/hostname and a port,
and internally it resolves to an IP address. This does not work for
network layers that rely on NAT64, like OpenThread.  Zephyr's
implementation of `getaddrinfo` is not aware of NAT64.  DNS will resolve
an IPV4 address that needs to be converted to IPV6 with the NAT64
prefix.

This commit alters the Hawkbit interface to allow providing an explicit
domain name as a string via `server_domain`, and an already resolved IP
address as `server_addr`.

This commit changes the usage of `hawkbit_runtime_config.server_addr` to
point to either an IP address or domain name. It adds a new Kconfig
(`HAWKBIT_USE_DOMAIN_NAME`) to specify an explicit domain name and adds
a new variable `hawkbit_runtime_config.server_domain`. If
`HAWKBIT_USE_DOMAIN_NAME` is enabled and a user provides an IP address
to `server_addr`, the user must provide a domain name to
`server_domain`.

Signed-off-by: Neal Jackson <neal@blueirislabs.com>

jukkar commented on include/zephyr/net/dns_resolve.h:
```diff
@@ -44,7 +44,7 @@ enum dns_query_type {
 
 /** Max size of the resolved name. */
 #ifndef DNS_MAX_NAME_SIZE
-#define DNS_MAX_NAME_SIZE 20
+#define DNS_MAX_NAME_SIZE 32
 #endif
```
Instead of hardcoding the value here, we could allow user to change the value via a Kconfig, so could you create a Kconfig option for this and set the default value to the present value (20).

jukkar commented on subsys/net/lib/sockets/Kconfig:
```diff
@@ -105,6 +105,7 @@ config NET_SOCKETS_SERVICE_THREAD_PRIO
 	int "Stack size for the thread handling socket services"
 	default 2400 if NET_DHCPV4_SERVER
+	default 1400 if DNS_RESOLVER
 	default 1400 if MDNS_RESPONDER
```
Please change this to
```
default 1400 if MDNS_RESPONDER || DNS_RESOLVER
```

nealjack replied:
This is complicated if DNS_MAX_NAME_SIZE is increased to an arbitrary value. 1400 may not be enough.
I would argue that DNS_MAX_NAME_SIZE should be 255 to be in line with [RFC1035](https://www.ietf.org/rfc/rfc1035.txt). I understand reducing memory usage for many cases in which domains are shorter, but I can see where this will be very confusing trying to get the right stack size for the net sockets service when using DNS and a dynamic DNS max length.

Perhaps the default should should be 255 and the stack size should be large enough to accommodate that, but users could adjust it to be a smaller value to reduce memory usage.

Any thoughts?

nealjack replied:
Actually, looking into it more, [`DNS_RESOLVER_MAX_QUERY_LEN`](https://github.com/zephyrproject-rtos/zephyr/blob/e0dbba5bd3ff18103fa3adb9c6efe2e44fcf278b/subsys/net/lib/dns/Kconfig#L64) seems to already be a Kconfig for this value, if I'm not mistaken. It's default is 255. I'm not sure why there is a redundant `DNS_MAX_NAME_SIZE` that is only 20 bytes.

jukkar replied:
We can certainly have a dedicated line for stack size if the 1400 is not enough.

But lets not increase the default value to such a high value, you are the first person that has reported this problem. So if it is possible to increase the max name len by the user via kconfig, then we are not introducing any changes how the system behaves by default.

jukkar replied:
> I'm not sure why there is a redundant `DNS_MAX_NAME_SIZE` that is only 20 bytes.

The other one is used in query, the other is where the response is stored. This has worked well so far, but if we have a Kconfig for setting `DNS_MAX_NAME_SIZE`, then user is certainly able to set those to same value.

nealjack replied:
Okay, so it seems the hawkbit subsystem has been misusing `DNS_MAX_NAME_SIZE`, as it is using it to size the string for the query. I propose replacing this with `DNS_RESOLVER_MAX_QUERY_LEN` where appropriate in the hawkbit code, and returning `DNS_MAX_NAME_SIZE` and the stack size configs to their original values.

Does this sound correct?

jukkar replied:
Yes, sounds about right.

nealjack replied:
This has been split to #89497

jukkar reviewed (commented):
Please remove the `subsys: ` prefix from commit subject of the both commits, it is not needed

maass-hamburg commented:
> This commit also changes how the `controllerId` is generated based on device id, and disentangles the two. The `controllerId` is what hawkbit uses to uniquely identify a device, and is not necessarily the same as the device id, and should be fully customizable by the user if needed. Previously, all custom device ids were being prepended with `CONFIG_BOARD`. When a user selects `CONFIG_HAWKBIT_CUSTOM_DEVICE_ID`, they should be able to specify the full `controllerId` used with hawkbit.

Please note, that changes that are not directly related to each other, should at least go into their own commit. 

This is even more unrelated and I recommend splitting it into it's own PR.

rlubos commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -778,12 +806,28 @@ int hawkbit_default_config_data_cb(const char *device_id, uint8_t *buffer, const
 				sizeof(hb_cfg.server_addr));
 			LOG_DBG("configured %s: %s", "hawkbit/server_addr", hb_cfg.server_addr);
 		}
+		if (config->server_hostname != NULL) {
```
That's an array so it can't be NULL, no?

nealjack replied:
`config` is of type [`hawkbit_runtime_config`](https://github.com/zephyrproject-rtos/zephyr/pull/89374/files/c776b00c6a0e4ebf8e3b9165e785adba1df8fca7#diff-ad25768e7f4edf2a111d7eaaf97f7aef126d77d3b4852c332706df975cdf7e04L31), and `server_hostname` is a `char *`.

The `config` structure is for the user to provide the string values to the hawkbit internals, where they are copied into the internal `hb_cfg`.

rlubos commented:
> For example changing the meaning of the setting "hawkbit/server_addr" and replacing it with "hawkbit/server_hostname" would lead to braking configurations of users that use that and then update to newest newest zephyr version. As the maintainer of the hawkbit subsystem, I'd like to not have such changes, if they are not necessary.

I agree we should avoid API changes, although I also see a need to be able to specify hostname when using numeric address representation for TLS purposes.

How about we simplify the change set, I. e. keep the current configuration as is (use `server_addr` for address resolving) to maintain backwards compatibility, but allow to specify `server_hostname` separately for TLS case, with a fallback to `server_addr` if not set?

nealjack commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -39,21 +39,21 @@
-#define DDI_SECURITY_TOKEN_SIZE 32
-#define RANGE_HEADER_SIZE 50
-#define HAWKBIT_RECV_TIMEOUT (300 * MSEC_PER_SEC)
+#define RECV_BUFFER_SIZE           640
```
The [coding style guidelines](https://docs.zephyrproject.org/latest/contribute/style/index.html#clang-format) requests that `clang-format` be run on contributions. I can revert and add another commit that runs the formatting?

maass-hamburg replied:
Where do you see that requirement?

> The [clang-format tool](https://clang.llvm.org/docs/ClangFormat.html) can be helpful to quickly reformat large amounts of new source code to our [Coding Style Guidelines](https://docs.zephyrproject.org/latest/contribute/style/index.html#coding-style-guidelines) standards together with the .clang-format configuration file provided in the repository.

It is a helper, but not mandatory. 

unrelated changes always must at least be in their own commit.

rlubos replied:
The link refers to the new code (and as already mentioned it's just a suggestion). Reformatting existing codebase in the same commit just blurs the diff.

nealjack commented:
> > This commit also changes how the `controllerId` is generated based on device id, and disentangles the two. The `controllerId` is what hawkbit uses to uniquely identify a device, and is not necessarily the same as the device id, and should be fully customizable by the user if needed. Previously, all custom device ids were being prepended with `CONFIG_BOARD`. When a user selects `CONFIG_HAWKBIT_CUSTOM_DEVICE_ID`, they should be able to specify the full `controllerId` used with hawkbit.
> 
> Please note, that changes that are not directly related to each other, should at least go into their own commit.
> 
> This is even more unrelated and I recommend splitting it into it's own PR.

I will work on splitting the PR to address this.

maass-hamburg commented:
> I agree we should avoid API changes, although I also see a need to be able to specify hostname when using numeric address representation for TLS purposes.

For that it's fine, but it should be put behind a Kconfig, which is disabled by default.

maass-hamburg commented:
> I could see a world where zephyr's implementation of getaddrinfo does the translation if using Openthread as the L2 layer, but as @rlubos mentions, it would result in a non-standard getaddrinfo implementation.

Put it behind a Kconfig, then it can be implemented and activated for only that use case.

nealjack commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -1265,8 +1306,7 @@ static void s_probe(void *o)
 
-	snprintk(url_buffer, sizeof(url_buffer), "%s/%s-%s", HAWKBIT_JSON_URL, CONFIG_BOARD,
-		 s->device_id);
+	snprintk(url_buffer, sizeof(url_buffer), "%s/%s", HAWKBIT_JSON_URL, s->controller_id);
```
This has been split to #89500

nealjack commented:
I split up this PR into three (#89500, #89497, and this one).

I have updated this PR to address specifically adding a `server_domain` field that can be enabled via Kconfig flag `HAWKBIT_USE_DOMAIN_NAME` which allows explicitly providing a domain name, and allowing `server_addr` to take an IP address instead of a domain name.

Yes, this is a workaround due to the Zephyr OS not providing support for NAT64 translation, especially with the dynamic NAT64 prefix generation that Openthread does. I don't think there was consensus about whether to alter `getaddrinfo`. I agree with @rlubos's reservations about altering a posix compliant implementation of getaddrinfo.

However, I am wondering if it would be more appropriate to alter Zephyr's DNS Resolve Library to be Openthread/NAT64 aware, as it is a zephyr-specific library above lower level posix calls. It allows users to provide a list of DNS servers at runtime and could be altered to accept a NAT64 prefix for translation at runtime too. This would require changing the hawkbit subsystem to use the async DNS resolve library.

I'm open to any ideas or thoughts.

nealjack commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -39,21 +39,21 @@
+#define RESPONSE_BUFFER_SIZE       1100
+#define DDI_SECURITY_TOKEN_SIZE    32
+#define RANGE_HEADER_SIZE          50
+#define HAWKBIT_RECV_TIMEOUT       (300 * MSEC_PER_SEC)
```
This has been undone in the most recent push

jukkar commented:
> However, I am wondering if it would be more appropriate to alter Zephyr's DNS Resolve Library to be Openthread/NAT64 aware, as it is a zephyr-specific library above lower level posix calls.

Note that getaddrinfo is using the zephyr dns resolver lib and not the other way around.

maass-hamburg commented:
@nealjack please check the compliance checks. you f.e. forgot the <> around your email in the signed off

rlubos commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -79,6 +79,9 @@ static struct hawkbit_config {
 #ifdef CONFIG_HAWKBIT_SET_SETTINGS_RUNTIME
 	char server_addr[DNS_MAX_NAME_SIZE + 1];
+#ifdef CONFIG_HAWKBIT_USE_DOMAIN_NAME
+	char server_domain[CONFIG_DNS_RESOLVER_MAX_QUERY_LEN + 1];
```
Same comment as in https://github.com/zephyrproject-rtos/zephyr/pull/89497#discussion_r2074960577 so it's not overlooked here. One option could be to define a new Kconfig for the buffer size for HawkBit, defaulting to DNS_RESOLVER_MAX_QUERY_LEN if DNS_RESOLVER is enabled.

nealjack replied:
What should be the default value if DNS_RESOLVER_MAX_QUERY_LEN is undefined?

For server_addr, the length can default to INET6_ADDRSTRLEN to support an IP address. However here it must support a DNS name. Is it okay to hardcode 255 as the default when there are no Kconfigs available?

for example:
```
#ifdef CONFIG_DNS_RESOLVER_MAX_QUERY_LEN
#define SERVER_DOMAIN_LEN CONFIG_DNS_RESOLVER_MAX_QUERY_LEN
#else¬
#define SERVER_DOMAIN_LEN 255
#endif¬
 ```

rlubos replied:
That's one option to have the default hardcoded as you've posted, or it can also be resolved via Kconfig if a dedicated config for the buffer lengths was created:
```
default DNS_RESOLVER_MAX_QUERY_LEN if DNS_RESOLVER 
default 255
```

nealjack replied:
I implemented a Kconfig for the domain name buffer length.

rlubos reviewed (commented):
Please fix the documentation

maass-hamburg commented:
@nealjack pls rebase, due to some bug that failed ci and is now fixed in the current main
